### PR TITLE
Layer and View

### PR DESCRIPTION
**Pull request by optionset**

### CODE REVIEW

File: src/Color.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
+
+export class Color {
+	public readonly red: number
+	public readonly green: number
+	public readonly blue: number
+	public readonly alpha: number
+
+	public constructor(red: number, green: number, blue: number, alpha: number) {
```
Comment (DaGrisa) [inline src/Color.ts]: check that the numbers are between 0 and 1

File: src/Color.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+	public static get red(): Color {
+		return Color._redColor
+	}
+
+	private static _transparentColor = new Color(0, 0, 0, 0.0)
+	public static get transparent(): Color {
+		return Color._transparentColor
+	}
+
+	// TODO: add some default colors (refer to UIColor)
```
Comment (DaGrisa) [inline src/Color.ts]: add when needed, or risk dead code

File: src/Geometry.ts [hunk 3]
```diff
@@ -0,0 +1,76 @@
+// TOOD: add method toString
+export class Point {
+	x: number
+	y: number
+
+	public constructor();
```
Comment (DaGrisa) [inline src/Geometry.ts]: why semicolon here?

File: src/Geometry.ts [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+	public constructor(x: number, y: number);
+	public constructor(other: Point);
```
Comment (DaGrisa) [inline src/Geometry.ts]: where is the implementation of these two constructors?

File: src/Geometry.ts [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+// TOOD: add method toString
+export class Rectangle {
+	x: number
+	y: number
+	width: number
+	height: number
+
+	public constructor()
+	public constructor(x: number, y: number, width: number, height: number);
+	public constructor(other: Rectangle);
```
Comment (DaGrisa) [inline src/Geometry.ts]: why semicolon here?

File: src/Geometry.ts (hunk 4 above)
Comment (optionset) [inline src/Geometry.ts]: TypeScript supports method overloading, but in the end you can only write one implementation of the method. There you have to take care of all the different implementations methods. That's the reason why the constructor below looks so weird:

`if`: copy constructor (line 8)
`else if`: explicit x and y (line 7)
`else`: defaults for x and y (line 6)

File: src/Geometry.ts (hunk 5 above)
Comment (optionset) [inline src/Geometry.ts]: semicolons: by accident
implementation: see above

File: src/Layer.ts [hunk 6]
```diff
@@ -0,0 +1,326 @@
+import { ArgumentException } from './ArgumentException'
+import { Color } from './Color'
+import { Point, Rectangle, Size } from './Geometry'
+
+export class Layer {
+	private __domElement: HTMLElement
```
Comment (DaGrisa) [inline src/Layer.ts]: why the double-underscore?

Comment (optionset) [inline src/Layer.ts]: only here a double underscore to show that we really don't want to use it too often.

File: src/Layer.ts [hunk 7]
```diff
@@ -0,0 +1,326 @@
...
+		this.backgroundColor = Color.transparent
+		this.borderColor = Color.black
+		this.borderWidth = 0.0
+		this.clipsToBounds = false
+		this.cornerRadius = 0.0
+		this.isHidden = false
+		this.opacity = 1.0
+		this.anchorPoint = new Point(0.5, 0.5)
+		this.frame = new Rectangle()
+	}
```
Comment (DaGrisa) [inline src/Layer.ts]: are the _ not needed here?

File: src/Layer.ts [hunk 8]
```diff
@@ -0,0 +1,326 @@
...
+	private _borderWidth: number
+	private _clipsToBounds: boolean
+	private _cornerRadius: number
+	private _isHidden: boolean
+	private _opacity: number
+
+	/**
+	 * The background color of the layer.
+	 *
+	 * The default value of this property is `null`
```
Comment (DaGrisa) [inline src/Layer.ts]: not null, transparent

File: src/Layer.ts (hunk 7 above)
Comment (optionset) [inline src/Layer.ts]: no, we want to invoke the property setter which will also set the appropriate styles and/or html attributes

File: src/Geometry.ts [hunk 9]
```diff
@@ -73,4 +42,81 @@ export class Size {
...
 	}
+
+	public toString(): string {
+		return `${this.width}x${this.height}`
+	}
+}
+
+export class Rectangle {
+	_position: Point
+	_size: Size
```
Comment (optionset) [inline src/Geometry.ts]: make `_position` and `_size` private
TypeScript‘s default access-level is public

Comment (optionset) [inline src/Geometry.ts]: if you need to access private or protected members in tests, you can take a look at the `getDomElement` helper in the Layer tests.

File: src/Geometry.ts [hunk 10]
```diff
@@ -1,4 +1,3 @@
-// TOOD: add method toString
 export class Point {
```
Comment (optionset) [inline src/Geometry.ts]: I‘m not sure if you should make the Geometry classes immutable — only getters, no setters. currently, there could be multiple references to one mutable rectangle and every one of these could change the shared instance. what do you think?

**Pull request merged**